### PR TITLE
MFLT-7240 Write documentation for Linux coredump feature

### PR DESCRIPTION
**Pull request by fnune**
Adds docs for the upcoming coredumps feature.

### CODE REVIEW

File: docs/ci/install-memfault-cli.mdx [hunk 1]
```diff
@@ -75,16 +76,17 @@ commands require which forms of authentication.
...
-| upload-mar                 |         ✅         |        ✅        |       ✅        |     ✅      |
 | upload-coredump            |         ✅         |        ✅        |       ✅        |     ✅      |
+| upload-mar                 |         ✅         |        ✅        |       ✅        |     ✅      |
 | upload-xed                 |         ✅         |        ✅        |       ✅        |     ✅      |
+| deploy-release             |         ✅         |        ✅        |       ✅        |             |
+| upload-android-app-symbols |         ✅         |        ✅        |       ✅        |             |
+| upload-aosp-symbols        |         ✅         |        ✅        |       ✅        |             |
 | upload-mcu-symbols         |         ✅         |        ✅        |       ✅        |             |
 | upload-ota-payload         |         ✅         |        ✅        |       ✅        |             |
+| upload-yocto-symbols       |         ✅         |        ✅        |       ✅        |             |
```
Comment (fnune) [inline docs/ci/install-memfault-cli.mdx:89]: Summary of changes in this section:

- I removed `Project Key` as a possible auth method of `deploy-release`
- Sorted the rows in order to group things together that have the same available auth methods
- Added `upload-yocto-symbols`

File: docs/linux/linux-coredumps.mdx [hunk 2]
```diff
@@ -0,0 +1,216 @@
+---
+id: linux-coredumps
+title: Linux Coredumps
+sidebar_label: Coredumps
+---
+
+import { ProjectMagicLink } from "@site/src/components/ProjectMagicLink";
+
+## Introduction
```
Comment (fnune) [inline docs/linux/linux-coredumps.mdx:9]: TODO: a screenshot would be nice.

File: docs/linux/linux-coredumps.mdx [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+Follow the [getting-started guide][docs-linux-getting-started] to learn how to
+set this up for your device. A key function of `memfaultd` is to preprocess and
+upload coredumps to the Memfault platform. It does this via `coredump_plugin`.
+
+`coredump_plugin` is enabled by default. Read more on [how to configure which
+plugins `memfaultd` builds with][docs-linux-control-plugins].
+
+### Linux kernel configuration
+
+Build your Linux kernel with the following options enabled:
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:60]: This makes it sound like there's an action to take, while I think it's highly likely to already be configured like this.

I'd rephrase to say something like

> Ensure that your Linux kernel with the following options enabled:


I asked [here](https://memfault.slack.com/archives/C037Z581MH9/p1664197171609859) how to check the existing defconfig/kconfig options:

Check `tmp/work/[machinename]/[kernelpackage]/[version]/build/.config` for your project's kernel configuration. For example: `tmp/work/qemuarm64-poky-linux/linux-yocto/5.15.62+gitAUTOINC+59c8898d45_7cb30c5e95-r0/linux-qemuarm64-standard-build/.config`

File: docs/linux/linux-coredumps.mdx [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+- `/proc/sys/kernel/core_pattern` (see [`man core`][man-core])
+- `/proc/sys/kernel/core_pipe_limit` (see [`man core`][man-core])
+
+Make sure your Linux image does not contain any other services that may write to
+these files. To do this, check for the inclusion of other coredump handlers
+(although uncommon in an embedded system, this may be Ubuntu's
+[Apport][apport]), and check your `/etc/sysctl.d` drop-in directory for anything
+other than Memfault that may be setting `kernel.core_pattern` or
+`kernel.core_pipe_limit`.
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:85]: Systemd also comes with coredump handling. That's more likely to be in the way than apport I think.

Looking at [this recipe for systemd](https://git.yoctoproject.org/poky/plain/meta/recipes-core/systemd/systemd_251.4.bb) there's a `coredump` option. Probably need to mention that it should not be used.

File: docs/linux/linux-coredumps.mdx [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+To check whether you've succeeded in letting `memfaultd` take care of these
+files, confirm that the contents of the `core_pattern` file reference the
+`memfault-core-handler` binary.
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:89]: The pattern is only written if memfaultd has data collection enabled. I'd mention that here.

File: docs/linux/linux-coredumps.mdx [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+## Uploading debugging symbols
+
+In order to process coredumps from your devices, you'll need to upload debugging
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:93]:
```suggestion
In order for Memfault to process coredumps from your devices, you'll need to upload debugging
```

File: docs/linux/linux-coredumps.mdx [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+credentials that you can use in your build environment or in continuous
+integration.
+
+Yocto's `do_rootfs` will produce a `.manifest` file (see [documentation on
+`IMAGE_MANIFEST`][yocto-image-manifest]) which lists all the installed packages
+that make up the image. The `memfault upload-yocto-symbols` commands makes use
+of it in order to find your build's debugging symbols.
+
+If you're not using Yocto, you can replicate its behavior by using our [REST API
+to upload symbols][api-docs-upload-symbols].
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:121]: OT: wonder if we should add a "generic" `upload-elf-symbols` command that takes a (directory with) ELF file(s), rather than pointing people to eject out of the CLI tool and use the REST API. Note this command would be a synonym for the current `upload-aosp-symbols` command.

File: docs/linux/linux-coredumps.mdx [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+```
+
+Read docs on [CI and authentication][docs-ci-auth] in order to obtain
+credentials that you can use in your build environment or in continuous
+integration.
+
+Yocto's `do_rootfs` will produce a `.manifest` file (see [documentation on
+`IMAGE_MANIFEST`][yocto-image-manifest]) which lists all the installed packages
+that make up the image. The `memfault upload-yocto-symbols` commands makes use
+of it in order to find your build's debugging symbols.
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:118]: In the help I had added:

>     The .manifest file is expected to reside in the location where Yocto produced the file.

to ensure hint that people shouldn't move the .manifest.

Do you think we should mention that here too?

File: docs/linux/linux-coredumps.mdx [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+<p>
+  You can always upload symbol files individually using the Memfault Web
+  Application. Open <b>Software -> Symbol Files</b> and click on{" "}
+  <b>Upload Symbol File</b>, or simply{" "}
+  <ProjectMagicLink path="/symbol-files?new">
+    click on this deep link to the app
+  </ProjectMagicLink>
+  .
+</p>
+
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:132]: People who've used Yocto + GDB may wonder whether they need to opt-in to any of the Yocto-provided debug options, like `IMAGE_GEN_DEBUGFS`, `DEBUG_BUILD`, `dbg-pkgs`, etc. Should we call out that none of these are needed and no additional changes to the device's image are required?

File: docs/linux/linux-coredumps.mdx [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+```
+
+To disable it:
+
+```shell
+$ memfaultd --disable-data-collection
+```
+
+The `memfaultd` service will restart automatically whenever you run either of
+those commands.
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:153]: Actually, only if it changed/toggled the setting:

https://github.com/memfault/memfault-linux-sdk-internal/blob/cffc3068dad3f4cadbe239cb4b29a5853f66d9f2/meta-memfault/recipes-memfault/memfaultd/files/memfaultd/src/memfaultd.c#L160-L161

File: docs/linux/linux-coredumps.mdx [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+Take a look at the [`/etc/memfaultd.conf` reference][docs-configuration] for
+more information.
+
+[source-builtin-conf]:
+  https://github.com/memfault/memfault-linux-sdk/blob/-/meta-memfault/recipes-memfault/memfaultd/files/memfaultd/builtin.conf
+
+## Configuring what's included in a coredump
+
+Currently, Memfault support for coredumps does not include custom filtering or
+selection of processes, memory regions or variables. However, the kernel does
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:164]: Should we mention that this will be added in a future SDK?

I'd rather not have people go all-in on `madvise` / etc. because those controls are super crude.
On top of that, once we add filtering ourselves, it will be counter-productive if those controls are *also* used by customers (will be harder to debug problems in the filtering logic).

File: docs/linux/linux-coredumps.mdx [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+## Testing your integration
+
+In order to test your integration, you'll need to make your device produce a
+coredump for a process, for example, by sending `SIGABRT` to it:
+
+```bash
+$ kill -s SIGABRT $YOUR_PROCESS_PID
+```
+
+Additionally, you might want to check that the `core_pattern` file is being set
```
Comment (martijnthe) [inline docs/linux/linux-coredumps.mdx:192]: Fine for now, but would probably move this piece to a "Troubleshooting" section on this page.

File: docs/linux/reference-memfaultd-configuration.mdx [hunk 13]
```diff
@@ -59,10 +65,11 @@ file.
...
-| `collectd_plugin`          | Configuration values for `plugin_collectd` if enabled in `memfaultd` (default). [Read more][docs-control-memfaultd-plugins].                                                                       | See [`reboot_plugin`](#reboot_plugin)     |
+| `swupdate_plugin`          | Configuration values for `plugin-swupdate` if enabled in `memfaultd` (default). [Read more][docs-control-memfaultd-plugins].                                                                       | See [`swupdate_plugin`](#swupdate_plugin) |
+| `reboot_plugin`            | Configuration values for `plugin-reboots` if enabled in `memfaultd` (default). [Read more][docs-control-memfaultd-plugins].                                                                        | See [`reboot_plugin`](#reboot_plugin)     |
+| `collectd_plugin`          | Configuration values for `plugin-collectd` if enabled in `memfaultd` (default). [Read more][docs-control-memfaultd-plugins].                                                                       | See [`reboot_plugin`](#reboot_plugin)     |
```
Comment (martijnthe) [inline docs/linux/reference-memfaultd-configuration.mdx:70]: Revert? The `_` was correct.

File: docs/linux/reference-memfaultd-configuration.mdx [hunk 14]
```diff
@@ -59,10 +65,11 @@ file.
...
 | `collectd`                 | Values used to configure CollectD.                                                                                                                                                                 | See [`collectd`](#collectd)               |
+| `coredump_plugin`          | Configuration values for `plugin-coredumps` if enabled in `memfaultd` (default). [Read more][docs-control-memfaultd-plugins].                                                                      | See [`coredump_plugin`](#coredump_plugin) |
```
Comment (martijnthe) [inline docs/linux/reference-memfaultd-configuration.mdx:72]:
```suggestion
| `coredump_plugin`          | Configuration values for `plugin_coredumps` if enabled in `memfaultd` (default). [Read more][docs-control-memfaultd-plugins].                                                                      | See [`coredump_plugin`](#coredump_plugin) |
```

File: docs/linux/reference-memfaultd-configuration.mdx [hunk 15]
```diff
@@ -126,6 +133,29 @@ file.
...
+}
+```
+
+To configure the location where coredumps are stored during processing, see
+[`data_dir`](#etcmemfaultdconf).
+
+| Field                         | Description                                                                                                        | Recommended value  |
+| ----------------------------- | ------------------------------------------------------------------------------------------------------------------ | ------------------ |
+| `rate_limit_count`            | The maximum amount of coredumps to process in a given period of `rate_limit_duration_seconds`.                     | 5                  |
+| `rate_limit_duration_seconds` | A window in which a maximum of `rate_limit_count` coredumps can be processed.                                      | 3600               |
```
Comment (martijnthe) [inline docs/linux/reference-memfaultd-configuration.mdx:155]: Should we add something like "Please don't change this setting without consulting Memfault"?

File: docs/linux/reference-memfaultd-configuration.mdx [hunk 16]
```diff
@@ -126,6 +133,29 @@ file.
...
+| `storage_max_usage_kib`       | A maximum of disk space to use for the purposes of coredump processing.                                            | _Project-specific_ |
+| `storage_min_headroom_kib`    | A minimum of disk space to leave unused in disk while processing a coredump. Honored over `storage_max_usage_kib`. | _Project-specific_ |
```
Comment (martijnthe) [inline docs/linux/reference-memfaultd-configuration.mdx:157]: > Honored over

Is it really "over"? Both are always honored.

File: docs/ci/install-memfault-cli.mdx [hunk 17]
```diff
@@ -75,16 +76,17 @@ commands require which forms of authentication.
...
-| deploy-release             |         ✅         |        ✅        |       ✅        |     ✅      |
 | post-chunk                 |                    |                  |                 |     ✅      |
-| upload-android-app-symbols |         ✅         |        ✅        |       ✅        |             |
-| upload-aosp-symbols        |         ✅         |        ✅        |       ✅        |             |
 | upload-bugreport           |         ✅         |        ✅        |       ✅        |     ✅      |
-| upload-mar                 |         ✅         |        ✅        |       ✅        |     ✅      |
 | upload-coredump            |         ✅         |        ✅        |       ✅        |     ✅      |
+| upload-mar                 |         ✅         |        ✅        |       ✅        |     ✅      |
 | upload-xed                 |         ✅         |        ✅        |       ✅        |     ✅      |
+| deploy-release             |         ✅         |        ✅        |       ✅        |             |
```
Comment (martijnthe) [inline docs/ci/install-memfault-cli.mdx:84]: nit: I'd keep the alphabetical ordering by command tbh

**Review by martijnthe — APPROVED** [review]
:heart:

File: docs/linux/linux-coredumps.mdx (hunk 3 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:60]: Awesome, thanks!

File: docs/linux/linux-coredumps.mdx (hunk 11 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:164]: Changed to:

```
Currently, Memfault support for coredumps does not include custom filtering or
selection of processes, memory regions or variables. These features are planned
for a future release.

However, the kernel does provide a certain degree of configurability. We'd
recommend waiting for a release of the Memfault Linux SDK with first-class
support for coredump filtering, but if your team needs it now:
```

File: docs/linux/reference-memfaultd-configuration.mdx (hunk 15 above)
Comment (fnune) [reply docs/linux/reference-memfaultd-configuration.mdx:155]: image
Added!

File: docs/linux/reference-memfaultd-configuration.mdx (hunk 16 above)
Comment (fnune) [reply docs/linux/reference-memfaultd-configuration.mdx:157]: Yah, removing that sentence.

Comment (fnune) [reply docs/linux/reference-memfaultd-configuration.mdx:157]: It's probably obvious

File: docs/ci/install-memfault-cli.mdx [hunk 18]
```diff
@@ -59,6 +59,7 @@ Commands:
   upload-ota-payload          Upload a binary to be used for an OTA update.
   upload-symbols              [DEPRECATED] Upload symbols for an MCU or...
   upload-xed                  Upload an .xed or .xcd file for analysis.
+  upload-yocto-symbols        Upload symbols for a Yocto build.
```
Comment (HBehrens) [inline docs/ci/install-memfault-cli.mdx:62]:
```suggestion
  upload-yocto-symbols        Upload symbols for a Linux Yocto build.
```

For symmetry reasons, the equivalent for Android is called "Android OS/AOSP".

File: docs/linux/introduction.mdx [hunk 19]
```diff
@@ -18,7 +18,7 @@ reason tracking, and in the future reporting crashes and logs.
 <p>
   <figure>
     <img
-      src="/img/docs/linux/sdk-overview-0.3.0.svg"
+      src="/img/docs/linux/sdk-overview-1.0.0.svg"
```
Comment (HBehrens) [inline docs/linux/introduction.mdx:21]: Does it really make sense to change that file name each time?

File: docs/linux/introduction.mdx [hunk 20]
```diff
@@ -51,26 +51,19 @@ If your project diverges from these assumptions, please [get in
...
-OTA/Release Management is currently fully supported through an off-the-shelf
-integration with the SWUpdate agent. Read more about it in the [OTA integration
-guide][docs-ota].
-
-[systemd-homepage]: https://systemd.io/
-[swupdate-homepage]: https://swupdate.org/
-[yocto-homepage]: https://www.yoctoproject.org/
-[docs-ota]: /docs/linux/linux-releases-integration-guide
-
 ## Documentation and Features
```
Comment (HBehrens) [inline docs/linux/introduction.mdx:63]: probably out of scope for this PR but it might make sene to offer glimpse at the UI for these respective features on this overview page.

File: docs/linux/introduction.mdx [hunk 21]
```diff
@@ -113,10 +106,21 @@ via [`statsd`][statsd-homepage].
...
+[Collect coredumps][docs-coredumps] and automatically upload them to the
+Memfault platform. Use Memfault's rich backtrace and trace analysis view in
+order to debug issues and keep your fleet's crashes under control. Make use of
+Memfault's deduplication features as well as advanced [Issue management and
+monitoring][docs-issue-management].
+
+[docs-coredumps]: /docs/linux/linux-coredumps
+[docs-issue-management]: /docs/platform/issue-management
+
+### Reboot Reason Tracking and Logs
```
Comment (HBehrens) [inline docs/linux/introduction.mdx:120]: Logs?

File: docs/linux/introduction.mdx [hunk 22]
```diff
@@ -113,10 +106,21 @@ via [`statsd`][statsd-homepage].
...
 
-These features are fully in the domain of `memfaultd`. Note that crash reports
-and logs are upcoming features, while reboot reason tracking is supported today.
+These features are fully in the domain of `memfaultd`. Note that logs are
+upcoming features, while reboot reason tracking is supported today.
```
Comment (HBehrens) [inline docs/linux/introduction.mdx:123]: nit: I'd flip the the order in this paragraph and talk about reboot reasons first (maybe with a brief example) and only closing with the outlook.

File: docs/linux/linux-coredumps.mdx [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+## Introduction
+
+From [`man core`][man-core]:
+
+> The default action of certain signals is to cause a process to terminate and
+> produce a core dump file, a file containing an image of the process's memory
+> at the time of termination. This image can be used in a debugger (e.g.,
+> gdb(1)) to inspect the state of the program at the time that it terminated. A
+> list of the signals which cause a process to dump core can be found in
+> signal(7).
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:18]: I'd move this section below our own intro paragraph.

File: docs/linux/linux-coredumps.mdx [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+
+By enabling coredump support with Memfault, the `memfaultd` daemon will
+automatically collect, preprocess and upload coredumps from your devices and
+send them to the Memfault platform. In turn, the Memfault platform acts as a
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:22]: send them immediately? regularly?

File: docs/linux/linux-coredumps.mdx [hunk 25]
```diff
@@ -0,0 +1,258 @@
...
+remote analyzer for your coredump files and (using your project's debugging
+symbols) is able to display a rich view of a coredump, displaying backtraces for
+all threads, and allowing you to inspect state in full detail.
+
+Additionally, Memfault takes care of [managing and deduplicating
+issues][docs-issue-management] as well as providing [metrics on
+issues][docs-issue-charts] and monitoring via
+[notifications][docs-notifications], granting you a clear image of how your
+fleet is behaving, as well as tight control over the success of your [OTA
+updates][docs-platform-ota].
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:32]: I like how we continue to push the docs towards describing the value gain beyond the plain feature description 👍

File: docs/linux/linux-coredumps.mdx [hunk 26]
```diff
@@ -0,0 +1,258 @@
...
+
+<ImageFigure
+  src="/img/docs/linux/linux-issue-detail-sleep-sigabrt.png"
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:35]: if we could have a screenshot that with more traces and devices, we could use the caption to hint at the value of fleet-wide coredump analysis, e.g. "A Linux coredump as it occurred hundreds of times already on 52 different devices in the past 24 hours. Memfault represents actionable details as a deduplicated Issue"

File: docs/linux/linux-coredumps.mdx [hunk 27]
```diff
@@ -0,0 +1,258 @@
...
+
+:::
+
+## Prerequisites
+
+### The `memfaultd` daemon, built with `coredump_plugin`
+
+Follow the [getting-started guide][docs-linux-getting-started] to learn how to
+set this up for your device. A key function of `memfaultd` is to preprocess and
+upload coredumps to the Memfault platform. It does this via `coredump_plugin`.
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:61]:
```suggestion
upload coredumps to the Memfault platform. It does this via its `coredump_plugin`.
```

File: docs/linux/linux-coredumps.mdx [hunk 28]
```diff
@@ -0,0 +1,258 @@
...
+### Linux kernel configuration
+
+Ensure that your Linux kernel is built with the following options enabled:
+
+- `CONFIG_COREDUMP=y` in order to enable coredump creation by the kernel
+- `CONFIG_CORE_DUMP_DEFAULT_ELF_HEADERS=y` in order to enable default mappings.
+  for processes. Specifically, Memfault requires:
+  - `bit 0`: dump anonymous private mappings,
+  - `bit 1`: dump anonymous shared mappings, and
+  - `bit 4`: (available since Linux 2.6.24) dump ELF headers.
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:75]: Where do these bits come from? Is this something the user can configure beyond a plain `y`? Is this configured elsewhere?

File: docs/linux/linux-coredumps.mdx [hunk 29]
```diff
@@ -0,0 +1,258 @@
...
+tmp/work/raspberrypi3-poky-linux/linux-yocto/5.15.62+gitAUTOINC+59c8898d45_7cb30c5e95-r0/linux-raspberrypi3-standard-build/.config`
+```
+
+Read [`man core`][man-core] in order to learn more about these settings. We
+recommend reading the entirety of this `man` page if you're in need of basic
+understanding of how coredumps work in Linux.
+
+### `core_pattern` and `core_pipe_limit`
+
+The following files need to be owned by `memfaultd` completely.
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:91]: The "owned" term sounds strange to me here. It unclear to me if these files normally exist or if they are also created by `memfaultd`

```suggestion
The following files will be modified by `memfaultd` at runtime and must not be written to by any other process.
```

File: docs/linux/linux-coredumps.mdx [hunk 30]
```diff
@@ -0,0 +1,258 @@
...
+
+Make sure your Linux image does not contain any other services that may write to
+these files. To do this, check for the inclusion of other coredump handlers such
+as SystemD (which can [act as a coredump handler when built with
+`-Dcoredump=true`][systemd-as-handler]), and check your `/etc/sysctl.d` drop-in
+directory for anything other than Memfault that may be setting
+`kernel.core_pattern` or `kernel.core_pipe_limit`.
+
+To check whether you've succeeded in letting `memfaultd` take care of these
+files, confirm that the contents of the `core_pattern` file reference the
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:104]: Do we later discuss how one confirms a correct setup at runtime? If so, maybe hint at this here?

File: docs/linux/linux-coredumps.mdx [hunk 31]
```diff
@@ -0,0 +1,258 @@
...
+`memfault-core-handler` binary.
+
+Note that `memfaultd` only sets `core_pattern`
+[if data collection is enabled](#set-enable_data_collection).
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:108]:
```suggestion
Note that `memfaultd` respects privacy settings only sets `core_pattern`
[if data collection is enabled](#set-enable_data_collection) (at runtime).
```

How do we handle the case when a user revokes data collection?

File: docs/linux/linux-coredumps.mdx [hunk 32]
```diff
@@ -0,0 +1,258 @@
...
+
+[systemd-as-handler]:
+  https://git.yoctoproject.org/poky/plain/meta/recipes-core/systemd/systemd_251.4.bb
+
+## Upload debugging symbols
+
+In order for Memfault to process coredumps from your devices, you'll need to
+upload debugging symbols for each of the binaries in your Linux image. We
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:116]: So far, the reader was purely thinking about software on the device. I am suggesting a few words here to ease the transition to the system-level thinking.

Ideally, we want a diagram here but I think that can wait.

```suggestion
After a coredump was collected on the device, the `memfaultd` will upload it to the Memfault Web Application. To allow the Memfault Web Application to reconstruct all details from such coredumps, you'll need to
upload debugging symbols for each of the binaries in your Linux image. We
```

File: docs/linux/linux-coredumps.mdx [hunk 33]
```diff
@@ -0,0 +1,258 @@
...
+recommend doing this as part of your build process (since this needs to be done
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:117]: The motivation is unclear, we need to do this for each new build of the software (hasn't been said so far). To reduce the risk of ever missing any software version, we recommend doing that in an automated fashion.

Maybe also worth calling out: It doesn't harm if you attempt to upload the same symbols multiple times. Memfault will detect this and the CLI will only upload what's new and truly needed.

File: docs/linux/linux-coredumps.mdx [hunk 34]
```diff
@@ -0,0 +1,258 @@
...
+Yocto's built-in debugging features such as `IMAGE_GEN_DEBUGFS`, `DEBUG_BUILD`
+or `dgb-pkgs`. The defaults can be used and will work with
+`memfault upload-yocto-symbols`.
+
+:::
+
+## Set `enable_data_collection`
+
+By default, `enable_data_collection` is `false` (see the [default
+configuration][source-builtin-conf]). This is to enable asking end users for
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:168]:
```suggestion
By default, `enable_data_collection` is `false` (see the [default
configuration][source-builtin-conf] to change this). This is to enable asking end users for
```

File: docs/linux/linux-coredumps.mdx [hunk 35]
```diff
@@ -0,0 +1,258 @@
...
+
+<p>
+  If your integration is all set and you've{" "}
+  <a href="#set-enable_data_collection">enabled data collection</a>, you'll be
+  able to see a new issue pop up in{" "}
+  <ProjectMagicLink path="/issues?live">
+    your project's Issues page
+  </ProjectMagicLink>
+  .
+</p>
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:244]: nit: why is this not Markdown?

File: docs/linux/linux-coredumps.mdx [hunk 36]
```diff
@@ -0,0 +1,258 @@
...
+correctly by `memfaultd` when it starts. Check that the output of this command
+includes `memfault-core-handler`:
+
+```bash
+$ cat /proc/sys/kernel/core_pattern
+```
+
+<p>
+  If your integration is all set and you've{" "}
+  <a href="#set-enable_data_collection">enabled data collection</a>, you'll be
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:238]: is there anything else they can already do on the device? e.g. logs of sorts?

File: docs/linux/linux-metrics.mdx [hunk 37]
```diff
@@ -212,7 +212,7 @@ $ memfaultd --disable-data-collection
 ```
 
 The `memfaultd` service will restart automatically whenever you run either of
-those commands.
+those commands and the value of the setting changes.
```
Comment (HBehrens) [inline docs/linux/linux-metrics.mdx:215]:
```suggestion
those commands if called with a value different from its current configuration.
```

File: docs/linux/reference-memfaultd-configuration.mdx [hunk 38]
```diff
@@ -126,6 +133,35 @@ file.
...
+| `rate_limit_count`<sup>\*</sup>            | The maximum amount of coredumps to process in a given period of `rate_limit_duration_seconds`. | 5                  |
+| `rate_limit_duration_seconds`<sup>\*</sup> | A window in which a maximum of `rate_limit_count` coredumps can be processed.                  | 3600               |
+| `storage_max_usage_kib`                    | A maximum of disk space to use for the purposes of coredump processing.                        | _Project-specific_ |
+| `storage_min_headroom_kib`                 | A minimum of disk space to leave unused in disk while processing a coredump.                   | _Project-specific_ |
+
+<p>
+  <sup>*</sup> Please{" "}
+  <a href="https://memfault.com/contact">consult with the Memfault team</a> if
+  you need to change rate-limiting settings for your integration.
+</p>
```
Comment (HBehrens) [inline docs/linux/reference-memfaultd-configuration.mdx:163]:
```suggestion
  <sup>*</sup> Please{" "}
  <a href="https://memfault.com/contact">consult with the Memfault team</a> if
  you need to change rate-limiting settings for your integration as the Memfault Web Application will further enforce [rate limiting rules](/docs/platform/rate-limiting).
</p>
```

Comment (HBehrens) [reply docs/linux/reference-memfaultd-configuration.mdx:163]: Do we need to update /docs/platform/rate-limiting with values for Linux?

**Review by HBehrens — APPROVED** [review]
Looks great, a couple of suggestions. Also, do we need to update `/docs/platform/rate-limiting`?

File: docs/linux/linux-coredumps.mdx (hunk 35 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:244]: Prettier fails to format `<ProjectMagicLink...` if I put it in the middle of the Markdown, removing necessary whitespace.

File: docs/ci/install-memfault-cli.mdx (hunk 18 above)
Comment (fnune) [reply docs/ci/install-memfault-cli.mdx:62]: This is the help message from the command. I agree, but I won't change it as part of this PR.

File: docs/linux/introduction.mdx (hunk 19 above)
Comment (fnune) [reply docs/linux/introduction.mdx:21]: :shrug: There's no harm here because no one needs to clone this repo for a core workflow when using Memfault (unlike the SDK, where I did keep just one file).

File: docs/linux/introduction.mdx (hunk 21 above)
Comment (fnune) [reply docs/linux/introduction.mdx:120]: I'll remove the mention of logs. This has been here from the start, though. I don't remember why we wanted it here.

File: docs/linux/introduction.mdx (hunk 22 above)
Comment (fnune) [reply docs/linux/introduction.mdx:123]: Part about `logs` removed

File: docs/linux/linux-coredumps.mdx (hunk 24 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:22]: IMO to early in the doc for this detail. Will consider mentioning this later.

File: docs/linux/linux-coredumps.mdx (hunk 28 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:75]: Two paragraphs later:

> Read [`man core`][man-core] in order to learn more about these settings. We
recommend reading the entirety of this `man` page if you're in need of basic
understanding of how coredumps work in Linux.

I'm moving that up to come right after this list.

File: docs/linux/linux-coredumps.mdx (hunk 30 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:104]: :+1: Added a link

File: docs/linux/linux-coredumps.mdx (hunk 31 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:108]: `journalctl` will log "won't post this corefile. Data collection disabled" or similar. I'll comment on `journalctl` logs later in the doc.

File: docs/linux/linux-coredumps.mdx (hunk 34 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:168]: They can't change it there. They would have to change it in their own `memfaultd.conf`, and IMO changing the default (setting it to `true`) shouldn't be recommended, since that would skip asking users for consent.

File: docs/linux/linux-coredumps.mdx (hunk 36 above)
Comment (fnune) [reply docs/linux/linux-coredumps.mdx:238]: Added a mention of `journalctl --unit memfaultd`.

File: docs/linux/reference-memfaultd-configuration.mdx (hunk 38 above)
Comment (fnune) [reply docs/linux/reference-memfaultd-configuration.mdx:163]: :+1: will update!

**Review by martijnthe — APPROVED** [review]

**Review by martijnthe — APPROVED** [review]

File: docs/linux/linux-coredumps.mdx [hunk 39]
```diff
@@ -124,26 +124,53 @@ part of continuous integration, in order to keep Memfault up to date with your
...
+  command-line arguments:
+</summary>
+
+- `--eu-unstrip-path`: path to a local `eu-unstrip` binary from
+  [elfutils][elfutils]. Note that one is available in
+  `tmp/sysroot-components/x86_64/elfutils-native` if you've added
+  `DEPENDS:append = " elfutils-native"` to your build.
+- `--package-debug-split-style`: your project's `PACKAGE_DEBUG_SPLIT_STYLE`. In
+  Poky, it defaults to `debug-with-srcpkg`. Read more about it [in the Yocto
+  reference][split-style].
```
Comment (HBehrens) [inline docs/linux/linux-coredumps.mdx:167]: this is super valuable detail, thank you!

**Review by HBehrens — APPROVED** [review]
LGTM
